lib/Service/Utils/FileUtils.php

```diff
@@ -0,0 +1,134 @@
+            $filename = $appendIdBeforeExtension($filename, $id);
+        }
+
+        // Ensure the filename length does not exceed the maximum allowed length.
+        $maxLength = 254;
+        if (mb_strlen($filename) > $maxLength) {
+            $pathInfo = pathinfo($filename);
+            $baseLength = $maxLength - mb_strlen($id) - 2; // Account for '_' and '.'.
+            if (isset($pathInfo['extension'])) {
+                $baseLength -= mb_strlen($pathInfo['extension']);
+                $filename = mb_substr($pathInfo['filename'], 0, $baseLength) . '_' . $id . '.' . $pathInfo['extension'];
+            } else {
+                $filename = mb_substr($filename, 0, $baseLength) . '_' . $id;
+            }
+        }
+
+        try {
+            // Validate the filename using the Nextcloud filename validator.
+            \OC::$server->get(\OCP\Files\IFilenameValidator::class)->validateFilename($filename);
+
+            // if recursion depth is greater than 0, log the change.
+            if ($recursionDepth > 0) {
+                $logger->info('Filename sanitized successfully: "' . $filename . '" (original: "' . $originalFilename . '")');
+            }
+
+            return $filename;
+        } catch (InvalidPathException $exception) {
+            $logger->warning('Invalid filename detected during sanitization: ' . $filename, ['exception' => $exception]);
+        }
+
+        // Handle specific exceptions and adjust the filename accordingly.
+        switch (true) {
+            case $exception instanceof FileNameTooLongException:
+                $filename = mb_substr($filename, 0, $maxLength - mb_strlen($id) - 2);
+                break;
+
+            case $exception instanceof EmptyFileNameException:
+                $filename = 'Untitled';
+                break;
+
+            case $exception instanceof InvalidCharacterInPathException:
+                if (preg_match('/"(.*?)"/', $exception->getMessage(), $matches)) {
+                    $invalidChars = array_merge(str_split($matches[1]), ['"']);
+                    $filename = str_replace($invalidChars, '-', $filename);
+                }
+                break;
+
+            case $exception instanceof InvalidDirectoryException:
+                $logger->error('Invalid directory detected in filename: ' . $exception->getMessage());
+                $filename = 'Untitled';
+                break;
+
+            case $exception instanceof ReservedWordException:
+                if (preg_match('/"(.*?)"/', $exception->getMessage(), $matches)) {
+                    $reservedWord = $matches[1];
+                    $filename = str_ireplace($reservedWord, '-' . $reservedWord . '-', $filename);
+                }
+                break;
+
+            default:
+                $logger->error('Unknown exception encountered during filename sanitization: ' . $filename);
+                $filename = 'Untitled';
+                break;
```

Comment:
Wow, this is a really elaborate. Thank you for the effort you put into this! I was actually thinking more of something like this: https://github.com/nextcloud/server/pull/51608/files#diff-911ea9939fad17c78ada50c38706874091e2478e37a2ef5a155481c0c4a81a98R144-R165 But it turns out the methods used there are not in OCP :(

Reply:
I'm inquiring now if we can get a proper sanitization method in OCP to avoid parsing error messages from the validator, which seems a bit brittle

Reply:
Here we go: https://github.com/nextcloud/server/pull/52688